src/Compilers/CSharp/Portable/Lowering/DiagnosticsPass_Warnings.cs

```diff
@@ -859,6 +860,50 @@ private bool CheckForAssignmentToSelf(BoundAssignmentOperator node)
             return false;
         }
 
+        private void CheckForDeconstructionAssignmentToSelf(BoundTupleLiteral leftTuple, BoundExpression right)
+        {
+            while (right.Kind == BoundKind.Conversion)
+            {
+
+                var conversion = (BoundConversion)right;
+                switch(conversion.ConversionKind)
+                {
+                    case ConversionKind.Deconstruction:
+                    case ConversionKind.ImplicitTupleLiteral:
+                    case ConversionKind.Identity:
+                        right = conversion.Operand;
+                        break;
+                    case var c:
+                        throw ExceptionUtilities.UnexpectedValue(c);
```

Comment:
>throw ExceptionUtilities.UnexpectedValue(c); [](start = 24, length = 44)

Are we certain we are not going to end up here even for erroneous code? Can we run into a user-defined conversion in non-error code?